radix-engine/src/model/access_controller/substates.rs

```diff
@@ -0,0 +1,57 @@
+use crate::types::*;
+use radix_engine_interface::time::Instant;
+
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct AccessControllerSubstate {
+    /// A vault where the asset controlled by the access controller lives.
+    pub controlled_asset: VaultId,
+
+    /// The amount of time (in minutes) that it takes for timed recovery to be done. Maximum is
+    /// 4,294,967,295 minutes which is 8171.5511700913 years. When this is [`None`], then timed
+    /// recovery can not be performed through this access controller.
+    pub timed_recovery_delay_in_minutes: Option<u32>,
+
+    /// The states of the Access Controller.
+    pub state: (PrimaryRoleState, OperationState),
```

Comment:
It would be more clear to have the P and R recovery states be separate so that the FSM is enforced by structure rather than relying on correct logic, e.g.

```
struct AccessControllerSubstate {
    ...
    pub state: (PrimaryRoleState, PState, RState),
    ...
}
    
pub enum PState {
    Normal,
    Recovery(Ruleset),
}

pub enum RState {
    Normal,
    Recovery(RRecovery),
}

pub enum RRecovery {
    Timed {
        pub timed_recovery_allowed_after: Instant,
        pub rule_set: RuleSet,
    },
    NoTimer(RuleSet),
}
```

Reply:
Sounds good, will split them up.